### PR TITLE
PR for 422

### PR DESCRIPTION
**Pull request by userquin**
Due to the fact client cannot download sha256.txt file (see https://github.com/AdoptOpenJDK/openjdk-website/issues/422#issuecomment-473655077), the changes are not finished, just need to parse the file and replace FILEHASH on the `autoResultCommand` text.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -212,6 +212,7 @@ gulp.task('browser-sync', () => {
         server: {
             baseDir: './'
         },
+        cors: true,
```
Comment (karianna) [inline gulpfile.js]: Security concern?

File: src/js/common.js [hunk 2]
```diff
@@ -153,12 +162,43 @@ module.exports.loadLatestAssets = (variant, openjdkImp, releaseType, release, ty
...
+
+module.exports.loadChecksum = (url, message) => {
+  return new Promise((resolve, reject) => {
+    loadUrl(url, response => {
+      try {
+        const reader = new FileReader();
+        // This fires after the blob has been read/loaded.
+        reader.addEventListener('loadend', e => {
+          // todo@userquin: fix this -> https://developer.mozilla.org/en-US/docs/Web/API/Event/srcElement
+          console.info(e.srcElement);
```
Comment (karianna) [inline src/js/common.js]: Would need to remove the console logging when done

File: src/js/common.js [hunk 3]
```diff
@@ -153,12 +162,43 @@ module.exports.loadLatestAssets = (variant, openjdkImp, releaseType, release, ty
...
 
+module.exports.getChecksum = url => checksums[url]
+
+module.exports.loadChecksum = (url, message) => {
+  return new Promise((resolve, reject) => {
+    loadUrl(url, response => {
+      try {
+        const reader = new FileReader();
+        // This fires after the blob has been read/loaded.
+        reader.addEventListener('loadend', e => {
```
Comment (karianna) [inline src/js/common.js]: What's an `e`?

File: src/js/common.js [hunk 4]
```diff
@@ -153,12 +162,43 @@ module.exports.loadLatestAssets = (variant, openjdkImp, releaseType, release, ty
...
+          console.info(e.target);
+          const text = e.srcElement.result;
+          const checksum = text.split(' ')[0].trim();
+          console.info(checksum);
+          checksums[url] = checksum;
+          resolve(message.replace('FILEHASH',checksum));
+        });
+        reader.readAsText(response.blob());
+      } catch (e) {
+        reject(e);
```
Comment (karianna) [inline src/js/common.js]: What happens to the user experience on a rejection?

File: gulpfile.js (hunk 1 above)
Comment (userquin) [reply gulpfile.js]: This is to allow access 302 redirection; if not included when connecting to download sha256.txt file, then the broswer will reject the request
On the other hand, if the response to the redirection from github does not includes cors headers on 302 response, the browser also reject the request

File: src/js/common.js (hunk 2 above)
Comment (userquin) [reply src/js/common.js]: yes, once tested this console will be removed

File: src/js/common.js (hunk 3 above)
Comment (userquin) [reply src/js/common.js]: e is the event once the blob is resolved: like another addEventListener callback

File: src/js/common.js (hunk 4 above)
Comment (userquin) [reply src/js/common.js]: in taht case the div container will not be shown, initially is hidden

**Comment by userquin** [comment]
> Made some comments but one of the more experienced web devs needs to chime in here. Also wondering if this is universally beneficial given its only windows.

My intention is to made it available to all platforms:

For linux/unix:
```shell
echo "FILEHASH  FILENAME" 2>&1 | sha256sum -c
```

and for macosx:
```shell
echo "FILEHASH FILENAME" 2>&1 | shasum -a 256 -c
```

File: src/json/config.schema.json [hunk 5]
```diff
@@ -142,6 +142,18 @@
             "title": "Checksum command",
             "examples": ["sha256sum FILENAME"]
           },
+          "checksumActionCommand": {
+            "$id": "#/properties/platforms/items/properties/checksumActionCommand",
+            "$ref": "#/definitions/command",
```
Comment (userquin) [inline src/json/config.schema.json]: the `#id` should be `#/properties/platforms/items/properties/checksumAutoCommand`

File: src/json/config.schema.json [hunk 6]
```diff
@@ -142,6 +142,18 @@
...
+            "title": "Checksum auto command",
+            "examples": ["sha256sum FILENAME"]
+          },
+          "checksumActionCommandResult": {
+            "$id": "#/properties/platforms/items/properties/checksumActionResultCommand",
+            "type": "string",
```
Comment (userquin) [inline src/json/config.schema.json]: the `#id` should be `#/properties/platforms/items/properties/checksumAutoResultCommand`

File: src/js/common.js (hunk 2 above)
Comment (userquin) [reply src/js/common.js]: In fact, I must check if responseText from xhr redirect response just returns the checksum line: in that case we just revert all changes made to handle xhr response on `commons.js` and here just remove the FileReader code...

File: src/js/common.js (hunk 4 above)
Comment (userquin) [reply src/js/common.js]: just see the staging server, there are errors downloading all sha256.txt files, and the div container is hidden, the page is the same as the original

**Comment by userquin** [comment]
@karianna the optimael solution will be if the server api includes the hash for each platform, this will prevent download each sha256.txt file and the auto comannd can be directly calculated.

The code is ready to handle this situation, and if the server api is changed to oncluded the hash, then 95% of the changes can be removed.

**Comment by gdams** [comment]
>  if the server api is changed to include the hash, then 95% of the changes can be removed.
@userquin I'm kind of pro doing work on the API to render the hash as a string and return it.

**Comment by userquin** [comment]
@gdams It is posible to include for each platform the checksum + checksum hash in the server api?

**Comment by gdams** [comment]
> @gdams It is posible to include for each platform the checksum + checksum hash algorithm in the server api response?

Yeah I can certainly take a look at that @userquin

**Comment by userquin** [comment]
@gdams any news on this?

**Comment by karianna** [comment]
@userquin Merge conflicts

**Comment by userquin** [comment]
This PR is not ready to be merged yet, I'm still waiting for @gdams responses. As I said, if the server api comes with hash and hash algh, we need only to adapt the ui, but if we need to call server to download each hash file, then some things must be changed on server side to avoid CORS errors.

**Comment by userquin** [comment]
Ok, it seems that client can now download hash files, so we need only update the client to read its content.

**Comment by userquin** [comment]
I'm at office, later at home I'll check it again, I'll try to add logic to parse the response, seems to be failing for same reason.

**Comment by userquin** [comment]
To solve the conflict (I dont know from what branch):
For current 2.0.1:
```javascript
// browser-sync task
gulp.task('browser-sync', function() {
    browserSync.init({
        server: {
            baseDir: './'
        },
        cors: true,
        notify: false
    });
});

// robots task
gulp.task('robots', function () {
  gulp.src('index.html')
    .pipe(robots({
      useragent: '*',
      allow: ['/'],
      disallow: ['/404.html', '/banner.html', '/dist']
    }))
    .pipe(gulp.dest('./'));
});
```
or for this conflict:
```javascript
// browser-sync task
gulp.task('browser-sync', function() {
    browserSync.init({
        server: {
            baseDir: './'
        },
        cors: true,
        notify: false
    });
});

// robots task
gulp.task('robots', function () {
  gulp.src('index.html')
    .pipe(robots({
      useragent: '*',
      allow: ['/'],
      disallow: ['/404.html', '/dist']
    }))
    .pipe(gulp.dest('./'));
});
```

**Comment by userquin** [comment]
@karianna The response for api data now has the cors header, but we also need cors header on responses for sha256.txt files:

api_call

sha256 txt

cors-errors

**Comment by userquin** [comment]
Well, by passing cors on chrome, just works, I need to do some final checks and I will update the PR. I need to check it on staging, but will not work due to CORS headers...

**Comment by userquin** [comment]
```javascript
2637dab3bc81274e19991eebc27684276b482dd71d0f84fedf703d4fba3576e5  OpenJDK8U-jdk_x64_windows_hotspot_8u202b08.zip
```
cors_disabled_chrome2

cors_disabled_chrome

**Comment by userquin** [comment]
@karianna I just push to my repo the changes and now it is working (see link below), and updated mac os x commands, I cannot test on linux now, maybe tomorrow  (it's 5 o'clock in the morning and I'm going to sleep;))..

I have tested on local on a Windows 10, starting chrome with no security: https://alfilatov.com/posts/run-chrome-without-cors/

File: src/js/common.js (hunk 2 above)
Comment (userquin) [reply src/js/common.js]: there is no need to use FileReader, just xhr.responseText works

@karianna to listen for events on html elements you add event listeners to them, provinding the event in question and a callback function: for example, a button can emit multiple events such as click, then we register an event listener to be notified when the user clicks on it:
```javascript
someButton.addEventListener("click", function(e) {
  // handle the event
}
```
The callback function can receive the event that fires the callback, you can call what the name you want. Using arrow functions, we can reduce the code removing extra code that is implicit:
```javascript
someButton.addEventListener("click", e => {
  // handle the event
}
```

**Comment by userquin** [comment]
Upps, there is something wrong: before update the master I just update but resolving the conflict updates changes @gdams do:

imagen

Some files was changed, but not present on my master branch (package.json and gulp.js), and at least gulp.js is wrong on my branch and in this PR (robots missing and robots task also missing on my gulp.js).

I dont want to push my changes, waiting to see where we are and how to solve the problem.

**Comment by userquin** [comment]
@gdams where are you touching? Your 2 last changes are missing from this master and I dont know where are included. See previous image.

**Comment by userquin** [comment]
@gdams Once I resolve the conflict manually on gulp.js file, I have compared my updated version againts your changes: as you can see bellow on the images, my gulp file has references to robots but clipboard no (copied from this master)

imagen

imagen

**Comment by userquin** [comment]
I'm trying another approach to avoid download the sha256.txt files for each platform, just using curl or wget  to download sha256 file and then check it, it is a good solution with only one command (for example, for mac):
```shell
curl -O -L FILEHASHURL && shasum -a 256 -c FILEHASHNAME
```
We can also hide download progress including `-s` flag:
```shell
curl -O -L -s FILEHASHURL && shasum -a 256 -c FILEHASHNAME
```
mac_checksum_alternative_command

**Comment by userquin** [comment]
For linux/unix (tested only on ubuntu 18):
```shell
wget -O- -T 1 -t 1 FILEHASHURL  | sha256sum -c
```
We can also hide download progress including `-q` flag:
```shell
wget -O- -q -T 1 -t 1 FILEHASHURL  | sha256sum -c
```
Captura de pantalla de 2019-04-18 16-21-40

**Comment by userquin** [comment]
And finally, after some tedious work, for windows (only `PowerShell >= 3.0`):

```shell
powershell -command "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12;  iwr -outf FILEHASHNAME FILEHASHURL" && powershell "$CHECKSUMVAR=($(Get-FileHash -Algorithm SHA256 -LiteralPath FILENAME | Format-List -Property Hash | Out-String) -replace \"Hash : \", \"\" -replace \"`r`n\", \"\"); Select-String -LiteralPath FILEHASHNAME -Pattern $CHECKSUMVAR | Format-List -Property FileName | Out-String" | find /i "FILEHASHNAME">Nul && ( echo "FILENAME: The SHA-256 fingerprint matches" ) || ( echo "FILENAME: The SHA-256 fingerprint does NOT match" )
```
windows_checksum_alternative_command

**Comment by userquin** [comment]
With this new approach, there is no need to download the sha256 files, so when the user executes the command (for ist corresponding SO), the sha256 file is downloaded and the file verified.
I'm updating the ui to reflect this changes.

**Comment by userquin** [comment]
@karianna, @gdams For windows I have not found a way to do it using windows shell, if some expert on it can rewrite the command instead using PowerShell, just need to replace the command on config.json file.

**Comment by karianna** [comment]
@gdams I'll await your review next.

**Comment by userquin** [comment]
@gdams can you ignore `package.json` and `package-lock.json`  from my PR? I dont know who and when touched these files and in which branch. I only updated my master branch  followed by a  `npm install` before making and pushing last changes in my master, and the result were those `package.json` and `package-lock.json` files.

**Comment by karianna** [comment]
@userquin I think you'll need to rebase/resolve the conflict and push again then I think this is good to go.

**Comment by userquin** [comment]
@karianna later at home I'll try or maybe tomorrow

**Comment by userquin** [comment]
@karianna there is no way to rebase package.json and package-lock.json from upstream/master, I'll try to just copy/paste both files into my master and try again...

**Comment by userquin** [comment]
ok, I just forget to do `npm install`, but then all versions change on package-lock.json and some dependencies were removed  (see images below).

just push or wait?

imagen

imagen
imagen

**Comment by karianna** [comment]
> ok, I just forget to do `npm install`, but then all versions change on package-lock.json and some dependencies were removed (see images below).
> 
> just push or wait?
> 
> imagen
> 
> imagen
> imagen

Yes push.

**Comment by userquin** [comment]
browser-sync fails on travis ci build, it needs browser-sync-client module to run, so I need to add it to `package.json`:
```json
    "browser-sync-client": "^2.26.4", <=== to fix build
    "browser-sync": "^2.26.5",
```

**Pull request merged**